File: androidApp/src/main/java/com/jaegerapps/trivitro_calculator/android/presentation/calculator_screen/CalculatorScreen.kt
package com.jaegerapps.trivitro_calculator.android.presentation.calculator_screen

import android.content.res.Configuration
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Scaffold
import androidx.compose.material.SnackbarHost
import androidx.compose.material.SnackbarHostState
import androidx.compose.material.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.jaegerapps.trivitro_calculator.android.R
import com.jaegerapps.trivitro_calculator.android.TrivitroTheme
import com.jaegerapps.trivitro_calculator.android.presentation.calculator_screen.components.*
import com.jaegerapps.trivitro_calculator.shared.presentation.calculator.CalculateUiEvent
import com.jaegerapps.trivitro_calculator.shared.presentation.calculator.CalculatorMode
import com.jaegerapps.trivitro_calculator.shared.presentation.calculator.CalculatorState
import com.jaegerapps.trivitro_calculator.shared.domain.models.PoolFilter
import kotlinx.coroutines.launch

@Composable
fun CalculatorScreen(
    calculatorState: CalculatorState,
    onCalculatorEvent: (CalculateUiEvent) -> Unit,
) {

    Surface(color = MaterialTheme.colors.surface) {
        when (calculatorState.mode) {
            CalculatorMode.BY_FILTER -> {
                ByFilterScreen(
                    poolFilterList = calculatorState.poolFilterList,
                    isChoosingManufacturer = calculatorState.isChoosingManufacturer,
                    isChoosingFilter = calculatorState.isChoosingFilter,
                    manufacturerText = calculatorState.manufacturerText,
                    filterText = calculatorState.filterText,
                    selectedFilter = calculatorState.selectedFilter,
                    onCalculatorEvent = {
                        onCalculatorEvent(it)
                    },
                )
            }

            CalculatorMode.BY_CUBIC_FEET -> {
                ByNumber(
                    displayText = calculatorState.input,
                    subtext = stringResource(R.string.subtext_cubic_feet),
                    selectedFilter = calculatorState.selectedFilter,
                    onCalculatorEvent = { onCalculatorEvent(it) }
                )
            }

            CalculatorMode.BY_SAND -> {
                ByNumber(
                    displayText = calculatorState.input,
                    subtext = stringResource(R.string.subtext_sand_needed),
                    selectedFilter = calculatorState.selectedFilter,
                    onCalculatorEvent = { onCalculatorEvent(it) }
                )
            }
        }
    }
}

@Composable
fun ByFilterScreen(
    poolFilterList: List<PoolFilter>,
    selectedFilter: PoolFilter? = null,
    manufacturerText: String? = null,
    filterText: String? = null,
    isChoosingFilter: Boolean = false,
    isChoosingManufacturer: Boolean = false,
    onCalculatorEvent: (CalculateUiEvent) -> Unit,
) {
    val snackbarHostState = remember { SnackbarHostState() }
    val scope = rememberCoroutineScope()
    val context = LocalContext.current

    Scaffold(
        snackbarHost = {
            SnackbarHost(hostState = snackbarHostState)
        },
        backgroundColor = MaterialTheme.colors.surface
    ) { innerPadding ->

        Column {
            TopBarCalc(
                onBackClick = { onCalculatorEvent(CalculateUiEvent.OnNavigate) },
                modifier = Modifier
                    .padding(innerPadding)
                    .shadow(5.dp)
            ) {
                DropDownBox(
                    list = poolFilterList.map { it.manufacturer }.distinct(),
                    currentItem = manufacturerText,
                    defaultText = stringResource(R.string.dropdown_select_manufacturer),
                    onEvent = {
                        onCalculatorEvent(CalculateUiEvent.SelectManufacturer(it))
                    },
                    onToggleEvent = { onCalculatorEvent(CalculateUiEvent.ToggleManufacturerDropdown) },
                    modifier = Modifier.fillMaxWidth(0.9f).testTag(stringResource(R.string.dropdown_select_manufacturer)),
                    showDropDown = isChoosingManufacturer
                )
                Spacer(Modifier.height(12.dp))
                DropDownBox(
                    list = poolFilterList.filter { it.manufacturer == manufacturerText }
                        .map { it.model },
                    currentItem = filterText,
                    defaultText = stringResource(R.string.dropdown_select_model),
                    onEvent = { filterString ->
                        if (!manufacturerText.isNullOrEmpty()) {
                            onCalculatorEvent(CalculateUiEvent.SelectFilter(filterString))
                        }
                    },
                    onToggleEvent = {
                        if (manufacturerText.isNullOrEmpty()) {
                            scope.launch {
                                snackbarHostState.showSnackbar(context.getString(R.string.prompt_select_manufacturer_first))
                            }
                        } else {
                            onCalculatorEvent(CalculateUiEvent.ToggleFilterDropdown)
                        }
                    },
                    modifier = Modifier
                        .fillMaxWidth(0.9f)
                        .testTag(stringResource(R.string.dropdown_select_model)),
                    showDropDown = isChoosingFilter
                )
            }
            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(horizontal = 12.dp),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                if (selectedFilter != null) {
                    StatDisplayColumn(selectedFilter)
                } else {
                    EmptyPrompt(
                        text = stringResource(id = R.string.prompt_start_calculating),
                        icon = com.jaegerapps.trivitro_calculator.R.drawable.icon_calculator
                    )
                }
            }
        }
    }
}

@Preview(showSystemUi = true)
@Composable
private fun ByFilterScreenPreview() {
    val selectedFilter = PoolFilter("Pentair", "TR40", 175, 140, 0, 0.0, 3.0)
    val exampleList = listOf(
        PoolFilter("Pentair", "TR40", 175, 140, 0, 0.0, 3.0),
        PoolFilter("Pentair", "TR50", 225, 180, 0, 0.0, 4.0),
        PoolFilter("Pentair", "TR60", 325, 260, 182, 78.0, 4.0),
        PoolFilter("Pentair", "TR60 ClearPro", 325, 260, 182, 78.0, 4.0),
        PoolFilter("Pentair", "TR100", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR100HD", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR100C-3", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR-140", 925, 740, 518, 222.0, 11.0),
        PoolFilter("Pentair", "TR140C-3", 925, 740, 518, 222.0, 11.0),
        PoolFilter("Hayward", "S144T", 50, 40, 40, 0.0, 1.0),
        PoolFilter("Hayward", "S160T", 150, 120, 120, 0.0, 3.0),
        PoolFilter("Waterway Carefree Top-Mount", "FS02629-B (26\")", 300, 240, 168, 72.0, 4.0)
    )
    TrivitroTheme {
        ByFilterScreen(
            poolFilterList = exampleList,
            isChoosingManufacturer = false,
            isChoosingFilter = true,
            manufacturerText = "Pentair",
            filterText = "TR40",
            selectedFilter = selectedFilter,
            onCalculatorEvent = {},
        )
    }
}
@Preview(showSystemUi = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun ByFilterScreenPreview_DARK() {
    val selectedFilter = PoolFilter("Pentair", "TR60", 325, 260, 182, 78.0, 4.0)

    val exampleList = listOf(
        PoolFilter("Pentair", "TR40", 175, 140, 0, 0.0, 3.0),
        PoolFilter("Pentair", "TR50", 225, 180, 0, 0.0, 4.0),
        PoolFilter("Pentair", "TR60", 325, 260, 182, 78.0, 4.0),
        PoolFilter("Pentair", "TR60 ClearPro", 325, 260, 182, 78.0, 4.0),
        PoolFilter("Pentair", "TR100", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR100HD", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR100C-3", 600, 480, 336, 144.0, 7.0),
        PoolFilter("Pentair", "TR-140", 925, 740, 518, 222.0, 11.0),
        PoolFilter("Pentair", "TR140C-3", 925, 740, 518, 222.0, 11.0),
        PoolFilter("Hayward", "S144T", 50, 40, 40, 0.0, 1.0),
        PoolFilter("Hayward", "S160T", 150, 120, 120, 0.0, 3.0),
        PoolFilter("Waterway Carefree Top-Mount", "FS02629-B (26\")", 300, 240, 168, 72.0, 4.0)
    )
    TrivitroTheme {
        ByFilterScreen(
            poolFilterList = exampleList,
            isChoosingManufacturer = false,
            isChoosingFilter = true,
            manufacturerText = "Pentair",
            filterText = "TR40",
            selectedFilter = selectedFilter,
            onCalculatorEvent = {},
        )
    }
}

@OptIn(ExperimentalComposeUiApi::class)
@Composable
fun ByNumber(
    displayText: String,
    subtext: String,
    selectedFilter: PoolFilter?,
    onCalculatorEvent: (CalculateUiEvent) -> Unit,
) {
    //ByNumber is used for all number cases. Inside the ViewModel, we have access to the mode
    //So in the ViewModel, we do calculations based on the mode, not here.
    val controller = LocalSoftwareKeyboardController.current

    Column {
        TopBarCalc(
            onBackClick = {
                controller?.hide()
                onCalculatorEvent(CalculateUiEvent.OnNavigate)
            },
            modifier = Modifier.shadow(5.dp)
        ) {
            NumberDisplay(
                text = displayText,
                subtext = subtext,
                modifier = Modifier.fillMaxWidth(0.9f),
                onClick = { onCalculatorEvent(it) },
                controller = controller
            )
        }
        Column(
            modifier = Modifier
                .weight(1f)
                .padding(horizontal = 12.dp),
            verticalArrangement = Arrangement.Center,
            horizontalAlignment = Alignment.CenterHorizontally,

        ) {
            if (selectedFilter != null) {
                StatDisplayColumn(selectedFilter)
            } else {
                EmptyPrompt(
                    text = stringResource(R.string.prompt_start_calculating),
                    icon = com.jaegerapps.trivitro_calculator.R.drawable.icon_calculator
                )
            }
        }
    }
}

@Preview(showSystemUi = true)
@Composable
private fun ByNumberPreview() {
    TrivitroTheme() {
        Surface(color = MaterialTheme.colors.surface) {
            ByNumber(
                displayText = "0",
                subtext = "sand needed",
                selectedFilter = null,
                onCalculatorEvent = {})
        }

    }
}
@Preview(showSystemUi = true, uiMode = Configuration.UI_MODE_NIGHT_YES)
@Composable
private fun ByNumberPreview_DARK() {
    TrivitroTheme() {
        Surface(color = MaterialTheme.colors.surface) {
            ByNumber(
                displayText = "0",
                subtext = "sand needed",
                selectedFilter = null,
                onCalculatorEvent = {})
        }

    }
}

File: iosApp/iosApp/main_screen/presentation/calculator/CalculatorScreen.swift
//
//  CalculatorScreen.swift
//  iosApp
//
//  Created by Hunter Freas on 2023/12/12.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

struct CalculatorScreen: View {
    
    let mode: String
    private var sharedState: SharedUiState
    @ObservedObject var calculatorVM: IosCalculatorViewModel
    
    init(sharedState: SharedUiState, mode: String) {
        self.sharedState = sharedState
        self.calculatorVM = IosCalculatorViewModel()
        self.mode = mode
       
    }
    
    var body: some View {
        VStack(alignment: .leading) {
            switch calculatorVM.state.mode {
            case .byFilter:
                ByFilter(
                    poolFilterList: calculatorVM.state.poolFilterList,
                    selectedFilter: calculatorVM.state.selectedFilter,
                    manufacturerText: calculatorVM.state.manufacturerText,
                    filterText: calculatorVM.state.filterText,
                    isChoosingManufacturer: calculatorVM.state.isChoosingManufacturer,
                    isChoosingModel: calculatorVM.state.isChoosingFilter,
                    onEvent: {item in calculatorVM.onEvent(event: item)}
                )
            case .byCubicFeet:
                ByNumber(
                    displayText: Binding(get: { calculatorVM.state.input }, set: { value in
                        calculatorVM.onEvent(event: CalculateUiEvent.OnNumberChange(value: value))
                    }),
                    subtext: "cubic feet",
                    selectedFilter: calculatorVM.state.selectedFilter,
                    onCalculatorEvent: { event in calculatorVM.onEvent(event: event) }
                )
            case .bySand:
                ByNumber(
                    displayText: Binding(get: { calculatorVM.state.input }, set: { value in
                        
                        calculatorVM.onEvent(event:CalculateUiEvent.OnNumberChange(value: value))
                    }),
                    subtext: "sand needed",
                    selectedFilter: calculatorVM.state.selectedFilter,
                    onCalculatorEvent: { event in calculatorVM.onEvent(event: event) }
                )
                
            default: Text(/*@START_MENU_TOKEN@*/"Hello, World!"/*@END_MENU_TOKEN@*/)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .onAppear {
            if !sharedState.poolFilterList.isEmpty {
                calculatorVM.onEvent(event: CalculateUiEvent.AddList(list: sharedState.poolFilterList))
            }
            calculatorVM.onEvent(event: CalculateUiEvent.ChangeMode(mode: mode))
            calculatorVM.startObserving()
        }
        .onDisappear {
            calculatorVM.dispose()
        }
        
            
        
    }
        
}

private struct ByFilter: View {
    
    let poolFilterList: [PoolFilter]
    let selectedFilter: PoolFilter?
    let manufacturerText: String?
    let filterText: String?
    let isChoosingManufacturer: Bool
    let isChoosingModel: Bool
    let onEvent: (CalculateUiEvent) -> Void
    
    @State private var toast: Toast? = nil
    
    var body: some View {
        VStack {
            TopBarCalc {
                VStack(spacing: 18) {
                    DropDownBox(
                        list: Array(poolFilterList.unique { $0.manufacturer}.map { $0.manufacturer}),
                        currentItem: manufacturerText,
                        showDropDown: isChoosingManufacturer,
                        defaultText: "Select manufacturer", onEvent: { item in onEvent(CalculateUiEvent.SelectManufacturer(text: item))},
                        onToggleEvent: {onEvent(CalculateUiEvent.ToggleManufacturerDropdown())}
                    ).accessibilityIdentifier("select manufacturer")
                    DropDownBox(
                        list: Array(poolFilterList.filter { $0.manufacturer == manufacturerText }.map { $0.model}),
                        currentItem: filterText,
                        showDropDown: isChoosingModel,
                        defaultText: "Select model",
                        onEvent: { item in onEvent(CalculateUiEvent.SelectFilter(filter: item))},
                        onToggleEvent: {
                            if manufacturerText == nil || manufacturerText == ""  {
                                toast = Toast(message: "Select a filter first.")
                            } else {
                                
                                    print("YO22")
                                onEvent(CalculateUiEvent.ToggleFilterDropdown())
                            }
                           
                        }
                    )
                    .accessibilityIdentifier("select model")
                }
                
            }
            VStack {
                if selectedFilter != nil {
                    StatDisplayColumn(selectedFilter: selectedFilter!, padding: 12)
                    
                } else {
                    Spacer()
                    EmptyPrompt(text: "Search for a filter", icon: "magnifyingglass")
                    Spacer()
                }
            }
           
        }
        .background(Color.surface)
        .toastView(toast: $toast)
        
        
    }
}

private struct ByNumber: View {
    @Binding var displayText: String
    let subtext: String
    let selectedFilter: PoolFilter?
    let onCalculatorEvent: (CalculateUiEvent) -> Void
    
    var body: some View {
        VStack {
            TopBarCalc {
                NumberDisplay(text: $displayText.wrappedValue, vmText: $displayText, subtext: subtext, onClick: { item in })
            }
            VStack {
                if selectedFilter != nil {
                    StatDisplayColumn(selectedFilter: selectedFilter!, padding: 12)
                    
                } else {
                    Spacer()
                    EmptyPrompt(text: "Start calculating now!", icon: "magnifyingglass")
                    Spacer()
                }
            }
        }
        .background(Color.surface)
    }
}



struct CalculatorScreen_ByFilter_Previews: PreviewProvider {
    static var previews: some View {
        CalculatorScreen(sharedState: SharedUiState(poolFilterList: [PoolFilter(manufacturer: "Pentair", model: "TR10", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR20", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR30", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR40", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR50", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR60", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR70", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5)], faqsList: [], isLoading: false, loaded: true, showOnboarding: false, error: nil), mode: "by_filter")
       
        
    }
}
struct CalculatorScreen_ByNumber_Previews: PreviewProvider {
    static var previews: some View {
        CalculatorScreen(sharedState: SharedUiState(poolFilterList: [PoolFilter(manufacturer: "Pentair", model: "TR10", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR20", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR30", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR40", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR50", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR60", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5),PoolFilter(manufacturer: "Pentair", model: "TR70", recommendedSandLoad: 1, recommendedVitrocleanVfaLoad: 2, recommendedPebble: 3, fiftyBagVitroclean: 4, fiftyBagPebble: 5)], faqsList: [], isLoading: false, loaded: true, showOnboarding: false, error: nil), mode: "by_sand")
       
        
    }
}
